vignettes/articles/glitter_for_Wikidata.Rmd

```diff
@@ -18,178 +18,225 @@ knitr::opts_chunk$set(
 
-```r
-stations_metro_Lyon=stations_metro_Lyon %>% 
-  transform_wikidata_coords("coords")
-```
+Now, we would like not only to view the stations but also the connecting lines. One property is of particular interest in this prospect: P197, which indicates **which other stations one station is connected to**. To form connecting lines, this information about the connection to other stations need to be complemented by *the involved line* and *direction* of that connection. Hence, we are not only interested in the **values** of the property P197, but also in the **property qualifiers** corresponding to the connecting line (P81) and direction (P5051)
 
-The resulting table may then be used easily with (for instance) package `leaflet`:
+We can thus complete our query this way:
 
-```r
-leaflet::leaflet(stations_metro_Lyon) %>%
-  leaflet::addTiles() %>%
-  leaflet::addCircles(popup=~stations_label)
+```{r query_prop_qualifiers}
+stations_adjacency=spq_init() %>% 
+  spq_add("?station wdt:P16 wd:Q1552") %>% 
+  spq_add("?station wdt:P31 wd:Q928830") %>%
+  spq_add("?station wdt:P625 ?coords") %>%
+  spq_add("?station p:P197 ?statement") %>%          # added instruction
+  spq_add("?statement ps:P197 ?adjacent") %>%        # added instruction
+  spq_add("?statement pq:P81 ?line") %>%             # added instruction
+  spq_add("?statement pq:P5051 ?direction") %>%      # added instruction
+  spq_label("station", "adjacent", "line", "direction",.overwrite=TRUE) %>% 
+  spq_select(-statement) %>% 
+  spq_perform() %>% 
+  na.omit() %>% 
+  select(coords,station,adjacent,line,direction)
+
+head(stations_adjacency)
 ```
 
+Now, we would like to put the stations **in the right order** so that we will be able to form the connecting lines. 
 
-# Example 2 : cities around Lyon
+This **data-wrangling part is a bit tricky** though not directly due to any glitter-related operation).
 
-Now, let's imagine that we are interested in the cities in a 200km radius around Lyon.
+We define a function `form_line()` which will put the rows in the table of stations in the correct order.
 
-```{r}
-find_item("Lyon")
-find_item("city")
+```{r form_line}
+form_line=function(adjacencies,direction){
+  N=nrow(adjacencies)
+  num=rep(NA,N)
+  ind=which(adjacencies$adjacent==direction)
+  i=N
+  num[ind]=i
+  while(i>1){
+    indnew=which(adjacencies$adjacent==adjacencies$station[ind])
+    ind=indnew
+    i=i-1
+    num[ind]=i
+  }
+  adjacencies=adjacencies %>% 
+    mutate(num=num) %>%
+    arrange(num) 
+  adjacencies=c(adjacencies$station,
+                direction)
+  return(adjacencies)
+}
 ```
 
-We could start exploring Wikidata with this query which finds all items that are instances ("wdt:P31") of "city" or of any subclass ("wdt:P279") of "city" . This query might return many items so that it seems reasonable to limit the number of items retrieved for now with the argument `limit`
+Now let's **apply this function to all lines and directions possible**:
```

Comment:
comment on how this uses the tidyverse?

Reply:
Yes, something like "Making full use of the tidyverse, we can use iteratively this function while not dropping the table-like structure of our data using a combination of tidyr::nest() and purrr::map()."